Use CreateFleet API when multiple instance types are specified in the compute resource configuration

enrico-usai opened the pull request:
## Description of changes

This patch permits to call boto3 `create-fleet` API when `InstanceTypeList` is specified in the cluster configuration.
The create-fleet response just contain the instance-id list, so we need to execute an additional describe-instances call
to retrieve other information like ip address to be saved in DynamoDB.

To pass the queue/compute-resource configuration parameters to the create-fleet api I'm using the `cluster-config.yaml` file.
This file contains the settings from the user and default values (e.g. `AllocationStrategy`).

The create-fleet api can return a list of Errors,
they will be logged as errors if there are no returned instances, as warning otherwise.

### New class hierarchy: `*FleetManager`
Within this patch we're moving responsibility to launch the EC2 instances from `InstanceManager` to `*FleetManager`.
The fleet manager will be selected (`Ec2InstanceManager` - `run-instances` API vs `Ec2FleetManager` - `create-fleet` API)
according to the configuration of the given compute resource.

### Run-instances-overrides
This patch renames internal variables from `run_instances_overrides` to `launch_overrides` to clarify
the code is able to pass the overridden parameters to both the APIs.
The configuration parameter and the file has not been renamed to preserve back-word compatibility.

### All or nothing
I'm adding a basic support for All-or-nothing feature, setting `MinTargetCapacity` to 1
when all_or_nothing is not configured.
If the minimum target capacity is not reached, the fleet launches no instances,
this might differ from run-instances approach where the run-instances call fails, to be verified.

### Multi-AZ support
In the create-fleet API we're explicitly setting `SingleAvailabilityZone` to `True`, the multi-AZ support is not included.

## Tests
Manually tested by updating the node package in a running cluster and reloading clustermgtd daemon
(slurm_resume changes don't require any reload).
To test this package I had to manually modify the `cluster-config.yaml` by adding `AllocationStrategy` and `InstanceTypeList` params
and adding some required policies to the Role attached to the head node.
Required policies are: `ec2:CreateFleet`, `ec2:RunInstances`, `ec2:DescribeCapacityReservations` and `ec2:CreateTags`.
* Manually tested config file with `SpotPrice: null` and with a value of 10. In the log I see `'Overrides': [{'MaxPrice': '10', 'InstanceType': 't2.medium'}]}],` if value is set, `'Overrides': [{'InstanceType': 't2.medium'}]}],` otherwise
* TODO: integration tests

## References
* https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.create_fleet

## Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
Use CreateFleet API when multiple instance types are specified in the compute resource configuration

This patch permits to call boto3 `create-fleet` API when `InstanceTypeList` is specified in the cluster configuration.
The create-fleet response just contain the instance-id list, so we need to execute an additional describe-instances call
to retrieve other information like ip address to be saved in DynamoDB.
Raises `Ec2InstancesInfoUnavailable` exception in `Ec2CreateFleetManager` too to signal the case when describe-instances call
returns incomplete info for 3 times (15 seconds).

To pass the queue/compute-resource configuration parameters to the create-fleet api I'm using the `cluster-config.yaml` file.
This file contains the settings from the user and default values (e.g. `AllocationStrategy`).

The create-fleet api can return a list of Errors,
they will be logged as errors if there are no returned instances, as warning otherwise.

## New class hierarchy: `*FleetManager`
Within this patch we're moving responsibility to launch the EC2 instances from `InstanceManager` to `*FleetManager`.
The fleet manager will be selected (`Ec2RunInstancesManager` - `run-instances` API vs `Ec2CreateFleetManager` - `create-fleet` API)
according to the configuration of the given compute resource.

## Run-instances-overrides
This patch renames internal variables from `run_instances_overrides` to `launch_overrides` to clarify
the code is able to pass the overridden parameters to both the APIs.
The configuration parameter and the file has not been renamed to preserve back-word compatibility.

## All or nothing
I'm adding a basic support for All-or-nothing feature, setting `MinTargetCapacity` to 1
when all_or_nothing is not configured.
If the minimum target capacity is not reached, the fleet launches no instances,
this might differ from run-instances approach where the run-instances call fails, to be verified.

## Multi-AZ support
In the create-fleet API we're explicitly setting `SingleAvailabilityZone` to `True`, the multi-AZ support is not included.

## Tests
Manually tested by updating the node package in a running cluster and reloading clustermgtd daemon
(slurm_resume changes don't require any reload).
To test this package I had to manually modify the `cluster-config.yaml` by adding `AllocationStrategy` and `InstanceTypeList` params
and adding some required policies to the Role attached to the head node.
Required policies are: `ec2:CreateFleet`, `ec2:RunInstances`, `ec2:DescribeCapacityReservations` and `ec2:CreateTags`.

### Unit tests
* Move test_instance_manager file outside slurm_resources folder.
* Move tests from test_instance_manager to test_fleet_manager.
* Add unit tests for *FleetManager classes, to verify create_fleet and describe_instances boto3 calls
* Mock launch_instances from fleet manager rather than boto3 calls in clustermgtd and resume unit tests.

## References
* https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.create_fleet

Signed-off-by: Enrico Usai <usai@amazon.com>

demartinofra commented on CHANGELOG.md:
```diff
@@ -3,6 +3,12 @@ aws-parallelcluster-node CHANGELOG
+------
+
+**ENHANCEMENTS**
+- Add support for flexible instance types, by calling create-fleet API when multiple instance types are specified in the cluster configuration.
```
nit: I would slightly reword this to capture the changes specific to this package. e.g. `Add support for EC2 Fleet as an alternative instance provisioning mechanism that allows greater flexibility in terms of instance diversifications and launch strategy`?

enrico-usai replied:
done

demartinofra commented on requirements.txt:
```diff
@@ -1,2 +1,3 @@
 boto3>=1.7.55
+PyYAML~=5.3
```
Shall we add a TODO to remove this once we change the config format or you'll remember to do so?

enrico-usai replied:
done. I'll left this comment open to be addressed when changing the config format.

demartinofra commented on src/slurm_plugin/clustermgtd.py:
```diff
@@ -218,6 +205,25 @@ def _get_launch_config(self, config):
         self.update_node_address = config.getboolean(
             "clustermgtd", "update_node_address", fallback=self.DEFAULTS.get("update_node_address")
         )
+        self.cluster_config_file = config.get(
+            "clustermgtd", "cluster_config_file", fallback=self.DEFAULTS.get("cluster_config_file")
+        )
```
reminder: add this to the __eq__ function when using the final config file.

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+class FleetManager(ABC):
+    """Abstract Fleet Manager."""
+
+    def __init__(
```
isn't this also supposed to be abstract so that FleetManager can't be instantiated?

optional: we might start adding typing info for the new methods we add.

enrico-usai replied:
done. Typing not completed

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+        Launch EC2 instances.
+
+        :raises ClientError in case of failures with Boto3 calls (run_instances, create_fleet, describe_instances)
+        :raises FleetManagerException in case of consecutive missing info
```
what do you mean by consecutive missing info? might be worth clarifying since it's not obvious from the implementation of launch_ec2_instances but requires digging into other calls

enrico-usai replied:
improved description.

demartinofra commented on src/slurm_plugin/slurm_resources.py:
```diff
@@ -63,6 +63,19 @@ def __str__(self):
         return hash(self.id)
 
+    @staticmethod
+    def from_info(instance_info):
```
nit: from_describe_instance_data/info?

enrico-usai replied:
changed

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+        return [EC2Instance.from_info(instance_info) for instance_info in assigned_nodes["Instances"]]
+
+
+class Ec2InstancesManager(FleetManager):
```
nit: would it make sense to keep the full EC2 api in the name: Ec2RunInstancesManager and Ec2CreateFleetManager?

enrico-usai replied:
done

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+        try:
+            response = create_fleet(self._region, self._boto3_config, launch_params)
+            for error in response.get("Errors", []):
+                logger.error("Error in CreateFleet request: %s - %s", error.get("ErrorCode"), error.get("ErrorMessage"))
```
is an error in the fleet response also always an error for our process? or is an error returned also when a subset of the instance types is not available and we don't want to mark it as an error?

enrico-usai replied:
This is something still to be verified/tested, from boto3 create_fleet documentation it is not clear.
This is a behaviour different from the standard boto3 calls on which a ClientError exception is raised.

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+
+            logger.debug("CreateFleet response: %s", response)
+            instance_ids = [inst_id for instances in response["Instances"] for inst_id in instances["InstanceIds"]]
+            return {"Instances": self._get_instances_info(instance_ids)}
```
in case we can't get the info for a single instance out of the entire batch it would be still ideal to avoid aborting the entire instance launch. Can we make it so that we return the instances we could describe and log an error for the missing ones that will be eventually terminated at the next iteration?

enrico-usai replied:
Done, removed the raising of the Exception.

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+        # Retrieve instance info from Instance Ids
+        instances = []
+        if instance_ids:
+            logger.info(f"Retrieving instance info for instances: {instance_ids}")
```
Shall we move this log to the _launch_instances function rephrasing to something like: launched the following EC2 instances:?

enrico-usai replied:
done

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+                except KeyError:
+                    failed_instance_id = instance_info["InstanceId"]
+                    partial_info_instances.append(failed_instance_id)
+                    instances.remove(instance_info)
```
you don't want to modify the collection while iterating over it, make a copy if you need to do so. Also a set is more efficient than a list in this case

enrico-usai replied:
You're right for the modification of the collection - fixed.
For the set 'm using a list because describe_instance_types requires a list at the end.

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+                    failed_instance_id = instance_info["InstanceId"]
+                    partial_info_instances.append(failed_instance_id)
+                    instances.remove(instance_info)
+                    logging.warning(f"Unable to retrieve all the required fields for instance {failed_instance_id}")
```
Can we log this only once for all failed instances?

enrico-usai replied:
Done

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+            if partial_info_instances:
+                # Try to find information for incomplete instances up to 3 times
+                if retry != 0:
+                    instances.extend(self._get_instances_info(partial_info_instances, retry - 1))
```
rather than calling this function recursively why don't we separate the logic in two functions, one that tries to retrieve the info and another one that manages the retry? it should make the entire code more readable by separating the responsibilities

enrico-usai replied:
Partially done. Let's talk about this.

enrico-usai replied:
Done by creating a private recursive function.

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,300 @@
+import boto3
+import yaml
+from botocore.exceptions import ClientError
+from slurm_plugin.slurm_resources import EC2Instance
```
optional: it would be nice to drop the dependency of fleet manager on something which is defined in slurm_plugin. I'd rather expect the other way around. Would it make sense to move EC2Instance out of the slurm modules?

enrico-usai replied:
Done, moved to fleet_manager.py module.

demartinofra reviewed (commented):
Nice work!

Commit:
Move EC2Instance from slurm_resources to fleet_manager module.

Fleet manager module should not depend on something which is defined in slurm_plugin.

Signed-off-by: Enrico Usai <usai@amazon.com>

enrico-usai commented on tests/slurm_plugin/test_clustermgtd.py:
```diff
@@ -34,7 +34,7 @@
 )
 
-from tests.common import MockedBoto3Request
+from tests.common import MockedBoto3Request, client_error, fleet_manager_factory
```
resolved.

enrico-usai commented on tests/slurm_plugin/test_fleet_manager.py:
```diff
@@ -0,0 +1,234 @@
+from assertpy import assert_that
+from slurm_plugin.fleet_manager import Ec2FleetManager, Ec2InstancesManager, FleetManagerFactory
+
+from tests.common import MockedBoto3Request, fleet_manager_factory
```
resolved

enrico-usai commented on tests/slurm_plugin/slurm_resources/test_instance_manager.py:
```diff
@@ -25,14 +26,14 @@
 )
 
-from tests.common import MockedBoto3Request
+from tests.common import MockedBoto3Request, client_error, fleet_manager_factory
```
resolved.

enrico-usai commented on tests/slurm_plugin/test_resume.py:
```diff
@@ -416,10 +345,26 @@
     )
-    # Only mock boto3 if testing case of valid clustermgtd heartbeat
+    # patch fleet manager
+    mock_fleet_manager = mocker.patch.object(
```
resolved.

lukeseawalker commented on src/slurm_plugin/instance_manager.py:
```diff
@@ -86,16 +88,23 @@ def add_instances_for_nodes(
+                )
+                launch_overrides = self._launch_overrides.get(queue, {}).get(compute_resource, {})
+
                 for batch_nodes in grouper(slurm_node_list, launch_batch_size):
```
does the `launch_batch_size`  has to be the same for both launch instance APIs? Can it be tuned differently or even removed?

enrico-usai replied:
It cannot be removed because we need to use a batch size when adding the items in route53. 
Yes in future we could have different size for the different apis.

lukeseawalker commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,386 @@
+            )
+        else:
+            raise Exception(
+                f"InstanceTypes or InstanceType field not found in compute resource config: {compute_resource_config}"
```
Do you need to print also queue name to be able to identify the compute resource uniquely?
I may be wrong but you can give two compute resources the same name under different queues

enrico-usai replied:
Changed. InstanceTypes vs InstanceTypeList too.

lukeseawalker commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,386 @@
+                else "spot",
+            },
+            "Type": "instant",
+            "SpotOptions": {
```
can be this set when launching only on-demand?

enrico-usai replied:
yes you can fill both of them and then select the one to use with `DefaultTargetCapacityType`

lukeseawalker commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,386 @@
+                # If the minimum target capacity is not reached, the fleet launches no instances
+                "MinTargetCapacity": 1 if not self._all_or_nothing else count,
+            },
+            "OnDemandOptions": {
```
same Q here, can be this set when launching only spot?

enrico-usai replied:
same here.

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,386 @@
+                # Try to find information for incomplete instances up to 3 times
+                message = f"Unable to retrieve instance info for instances: {partial_instance_ids}"
+                if retry != 0:
+                    logging.warning(message + ". Retrying...")
```
Use logger (same in other places)

enrico-usai replied:
done.

demartinofra commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,386 @@
+        :raises ClientError in case of boto3 failure
+        """
+
+        def _get_instances_info_or_retry(instance_ids_to_retry: list, retry=3):
```
in the end this private function includes everything so at this point you could even get rid of it and assign retry to a local variable.

What I initially had in mind was something like this:
```
    # in _launch_instances
    instance_ids = [inst_id for instance in instances for inst_id in instance["InstanceIds"]]
    complete_instances, partial_instance_ids = self._get_instances_info(instance_ids)
    if partial_instance_ids:
        message = f"Unable to retrieve instance info for instances: {partial_instance_ids}"
        logging.error(message)
    return {"Instances": complete_instances}


    def _get_instances_info(self, instance_ids: list):
        """
        Describe instances to retrieve info not available from create-fleet response.
        :raises ClientError in case of boto3 failure
        """
        complete_instances = []
        partial_instance_ids = list(instance_ids)
        
        
        retry = 3

        while retry > 0 and partial_instance_ids:
            # Wait for instances to be available in EC2
            time.sleep(5)
            result, partial_instance_ids = _retrieve_instances_info_from_ec2(partial_instance_ids)
            complete_instances.append(result)
            retry = retry-1

        return complete_instances, partial_instance_ids
        

    def _retrieve_instances_info_from_ec2(self, instance_ids: list):
        """Retrieve instance info from EC2 by Instance Ids."""
        complete_instances = []
        partial_info_instances = []
        if instance_ids:
            ec2_client = boto3.client("ec2", region_name=self._region, config=self._boto3_config)
            paginator = ec2_client.get_paginator("describe_instances")
            response_iterator = paginator.paginate(InstanceIds=instance_ids)
            filtered_iterator = response_iterator.search("Reservations[].Instances[]")
            instances = [instance_info for instance_info in filtered_iterator]
            logging.debug("Launched instances: %s", instances)

            for instance_info in filtered_iterator
                try:
                    # Try to build EC2Instance objects using all the required fields
                    EC2Instance.from_describe_instance_data(instance_info)
                    complete_instances.append(instance_info)
                except KeyError:
                    partial_info_instances.append(instance_info["InstanceId"])

        return complete_instances, partial_info_instances
```

enrico-usai replied:
Added your proposal, except for `extend` vs `append`.

lukeseawalker commented on tests/slurm_plugin/test_clustermgtd.py:
```diff
@@ -912,12 +917,16 @@ def test_handle_unhealthy_static_nodes(
         }
     )
-    cluster_manager._instance_manager._launch_ec2_instances = mocker.MagicMock(return_value=launched_instances)
+    # cluster_manager._instance_manager._launch_ec2_instances = mocker.MagicMock(return_value=launched_instances)
```
remove if not needed

enrico-usai replied:
done

lukeseawalker commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,383 @@
+                        "LaunchTemplateName": f"{self._cluster_name}-{self._queue}-{self._compute_resource}",
+                        "Version": "$Latest",
+                    },
+                    "Overrides": overrides,
```
missing the `LaunchTemplateConfigs/Overrides/MaxPrice` for the spot case. Correspond to the ParallelCluster configuration option [SpotPrice](https://docs.aws.amazon.com/parallelcluster/latest/ug/Scheduling-v3.html#yaml-Scheduling-SlurmQueues-ComputeResources-SpotPrice)

enrico-usai replied:
Added. 

Manually tested config file with `SpotPrice: null` and with a value of 10. In the log I see `'Overrides': [{'MaxPrice': '10', 'InstanceType': 't2.medium'}]}],` if value is set, `'Overrides': [{'InstanceType': 't2.medium'}]}],` otherwise

lukeseawalker commented on src/slurm_plugin/fleet_manager.py:
```diff
@@ -0,0 +1,383 @@
+                if self._queue_config["CapacityType"] == "ONDEMAND"
+                else "spot",
+            },
+            "Type": "instant",
```
missing `TagSpecifications`

demartinofra reviewed (approved):
🚀🚀🚀🚀🚀
